apps/tangle-dapp/containers/ManageProfileModalContainer/AllocationInput.tsx

```diff
@@ -0,0 +1,298 @@
+import { BN } from '@polkadot/util';
+import {
+  ChevronDown,
+  ChevronUp,
+  Close,
+  LockUnlockLineIcon,
+} from '@webb-tools/icons';
+import { IconBase, IconSize } from '@webb-tools/icons/types';
+import {
+  Chip,
+  IconWithTooltip,
+  Input,
+  InputWrapper,
+  Label,
+  SkeletonLoader,
+  Typography,
+} from '@webb-tools/webb-ui-components';
+import { FC, ReactElement, useCallback, useState } from 'react';
+import { twMerge } from 'tailwind-merge';
+import { z } from 'zod';
+
+import { TANGLE_TOKEN_UNIT } from '../../constants';
+import usePolkadotApiRx from '../../hooks/usePolkadotApiRx';
+import { ServiceType } from '../../types';
+import convertAmountStringToChainUnits from '../../utils/convertAmountStringToChainUnits';
+import convertChainUnitsToNumber from '../../utils/convertChainUnitsToNumber';
+import { formatTokenBalance } from '../../utils/polkadot/tokens';
+
+export type AllocationInputProps = {
+  amount: BN | null;
+  availableServices: ServiceType[];
+  availableBalance: BN | null;
+  service: ServiceType | null;
+  validateAmountAgainstRemaining: boolean;
+  title: string;
+  id: string;
+  isDisabled?: boolean;
+  hasDeleteButton?: boolean;
+  isLocked?: boolean;
+  lockTooltip?: string;
+  onDelete?: (service: ServiceType) => void;
+  onChange?: (newAmountInChainUnits: BN) => void;
+  setService: (service: ServiceType) => void;
+};
+
+export function getRoleChipColor(
+  role: ServiceType
+): 'green' | 'blue' | 'purple' {
+  switch (role) {
+    case ServiceType.ZK_SAAS_GROTH16:
+    case ServiceType.ZK_SAAS_MARLIN:
+      return 'blue';
+    case ServiceType.TX_RELAY:
+      return 'green';
+    case ServiceType.DKG_TSS_CGGMP:
+      return 'purple';
+  }
+}
+
+const DECIMAL_REGEX = /^\d*(\.\d+)?$/;
+
+const STATIC_VALIDATION_SCHEMA = z
+  .string()
+  .regex(
+    DECIMAL_REGEX,
+    'Only digits or numbers with a decimal point are allowed'
+  );
+
+const AllocationInput: FC<AllocationInputProps> = ({
+  isLocked = false,
+  lockTooltip,
+  amount = null,
+  availableBalance,
+  hasDeleteButton = false,
+  isDisabled = false,
+  validateAmountAgainstRemaining,
+  availableServices,
+  title,
+  id,
+  service,
+  onChange,
+  setService,
+  onDelete,
+}) => {
+  const [isDropdownVisible, setIsDropdownVisible] = useState(false);
+
+  const { data: minRestakingBond } = usePolkadotApiRx(
+    useCallback((api) => api.query.roles.minRestakingBond(), [])
+  );
+
+  const handleAmountChange = useCallback(
+    (newValue: string) => {
+      // Do nothing if the input is invalid or empty.
+      if (
+        onChange === undefined ||
+        newValue === '' ||
+        !DECIMAL_REGEX.test(newValue)
+      ) {
+        return;
+      }
+
+      const newAmountInChainUnits = convertAmountStringToChainUnits(newValue);
+
+      onChange(newAmountInChainUnits);
+    },
+    [onChange]
+  );
+
+  const handleDelete = () => {
```

Comment:
I think it's best if we add `useMemo` and `useCallback` in variables and functions below

Reply:
One of the functions I didn't add `useCallback` for, because it's not passed directly to a child as a prop, instead it needs to be wrapped in a closure `() => ...`, and I've analyzed & determined that adding `useCallback` here might not be worth the complexity added to properly accomplish that. In case that we do run into performance issues in the future, then I'll revisit it.